### PR TITLE
feat: add flow support

### PR DESCRIPTION
**Pull request by moroine**

### CODE REVIEW

File: __tests__/helpers.js [hunk 1]
```diff
@@ -26,90 +27,90 @@ describe('helpers', () => {
...
-          expect(helpers.parseISO(date)).to.deep.equal(new Date(date));
+          expect(helpers.parseISO(date)).toEqual(new Date(date));
         });
       });
 
       describe('date is a integer type', () => {
         it('should return date as it is', () => {
           const date = 1514764800000; // 2018-01-01
-          expect(helpers.parseISO(date)).to.equal(date);
+          expect(helpers.parseISO(date)).toEqual(new Date(date));
```
Comment (limtingzhi) [inline __tests__/helpers.js:44]: The result changed? Why though? 

The `it` description needs to change too.

File: src/legacy.js [hunk 2]
```diff
@@ -1,7 +1,9 @@
+// @flow
+
 import { getAutoCompareRangeAndLabel, getCompareRange, getCustomPeriod, getPeriodParams, getRange } from './main';
+import type { CompareMode, DateLike, LegacyCompareMode, LegacyPeriod, Period, PeriodParams, Range } from './types.js.flow';
 
-// (Object | String) -> String
-const migrateLegacyPeriod = (period) => {
+const migrateLegacyPeriod = (period: Period | Range): Period => {
```
Comment (limtingzhi) [inline src/legacy.js:6]: `Period | Range` can be replaced by `LegacyPeriod`?

File: src/legacy.js [hunk 3]
```diff
@@ -64,20 +59,31 @@ const toLegacyCompareMode = (compareMode) => {
...
 
-// :: (String | Object) -> Option(Date | Int | String) = new Date() -> Object
-const calculateAutoCompare = (period, base) => {
+type CalculateAutoCompareReturn = {|
+  label: string,
+  period: Range,
+|};
+
+const calculateAutoCompare = (
+  period: LegacyPeriod, base?: Date = new Date(),
```
Comment (limtingzhi) [inline src/legacy.js:85]: Should `base` be `DateLike` instead?

File: src/main.js [hunk 4]
```diff
@@ -1,23 +1,25 @@
...
-// :: (Date | Int | String) -> (Date | Int | String)  -> Object
-const formatRange = (start, end) => ({
+const formatRange = (start: DateLike, end: DateLike): Range => ({
   start: formatDate(start),
   end: formatDate(end),
 });
 
-// :: Int -> String -> Option(Date | Int | String) = new Date() -> Object
-const getLastRelativePeriodRange = (num, unit, base = new Date()) => {
+const getLastRelativePeriodRange = (num: number, unit: Unit, base: Date = new Date()): Range => {
```
Comment (limtingzhi) [inline src/main.js:12]: Add `?` to `base` as it is optional?

And should `base` be `DateLike` instead?

File: src/main.js [hunk 5]
```diff
@@ -1,23 +1,25 @@
...
   const endOf = getEndOfFn(unit);
   const sub = getSubtractionFn(unit);
 
   return formatRange(startOf(sub(base, num)), endOf(sub(base, 1)));
 };
 
-// :: String -> Option(Date | Int | String) = new Date() -> Object
-const getThisRelativePeriodRange = (unit, base = new Date(), num = 1) => {
+const getThisRelativePeriodRange = (
+  unit: Unit, base: Date = new Date(), num: number = 1,
```
Comment (limtingzhi) [inline src/main.js:21]: Add `?` to `base` and `num` as they are optional?

And should `base` be `DateLike` instead?

File: src/main.js [hunk 6]
```diff
@@ -26,15 +28,26 @@ const getThisRelativePeriodRange = (unit, base = new Date(), num = 1) => {
   return formatRange(startOf(sub(dayBefore, num - 1)), formatDate(dayBefore));
 };
 
-// :: (Date | Int | String) -> Option(Date | Int | String) = new Date() -> Object
-const getTillYesterdayRange = (start, base = new Date()) => {
+const getTillYesterdayRange = (start: DateLike, base: Date = new Date()): Range => {
```
Comment (limtingzhi) [inline src/main.js:31]: Add `?` to `base` as it is optional?

And should `base` be `DateLike` instead?

File: src/main.js [hunk 7]
```diff
@@ -82,8 +95,9 @@ const getPeriodParams = (period) => {
   }
 };
 
-// :: (String) -> Option(Date | Int | String) -> Option(Int) = 0 -> Object
-const getRange = (period, base = new Date(), utcOffset = 0) => {
+const getRange = (
+  period: Period, base: DateLike = new Date(), utcOffset: string | number = 0,
```
Comment (limtingzhi) [inline src/main.js:99]: Add `?` to `base` and `utcOffset` as they are optional?

File: src/main.js [hunk 8]
```diff
@@ -140,8 +153,9 @@ const getBestCompareUnit = (start, end) => {
   return 'day';
 };
 
-// :: (String) -> Option(Date | Int | String) = new Date() -> Object
-const getAutoCompareRangeAndLabel = (period, baseDate = new Date()) => {
+const getAutoCompareRangeAndLabel = (
+  period: Period, baseDate: Date = new Date(),
```
Comment (limtingzhi) [inline src/main.js:157]: Add `?` to `baseDate` as it is optional?

And should `baseDate` be `DateLike` instead?

File: src/main.js [hunk 9]
```diff
@@ -190,8 +204,7 @@ const getAutoCompareRangeAndLabel = (period, baseDate = new Date()) => {
   }
 };
 
-// :: String -> Option(String) = 'auto' -> Object
-const getCompareRange = (period, compareMode = 'auto') => {
+const getCompareRange = (period: Period, compareMode: CompareMode = 'auto'): Range => {
```
Comment (limtingzhi) [inline src/main.js:207]: Add `?` to `compareMode` as it is optional?

File: src/types.js.flow [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+export type PeriodYesterdayParams = {|
+  type: 'yesterday',
+|};
+
+export type PeriodAllTimeParams = {|
+  type: 'all_time',
+|};
+
+export type PeriodParams = PeriodLastParams | PeriodThisParams | PeriodTillYesterdayParams
+| PeriodCustomParams | PeriodYesterdayParams | PeriodAllTimeParams;
```
Comment (limtingzhi) [inline src/types.js.flow:45]: Could you indent this?

File: __tests__/helpers.js (hunk 1 above)
Comment (moroine) [reply __tests__/helpers.js:44]: Because `parseISO` is supposed to convert to `Date` right? Previously timestamp was kept untouched

Comment (limtingzhi) [reply __tests__/helpers.js:44]: Oh okay, could you update the `it` description to `should return date as date type`?

**Review by limtingzhi — APPROVED** [review]

**Pull request merged**